### PR TITLE
Intel txt aem rebased

### PR DESCRIPTION
**Pull request by tomzy-0**
This PR includes changes needed for Intel TXT AEM support rebased on top of GRUB 2.06 release.

### CODE REVIEW

File: grub-core/loader/i386/linux.c [hunk 1]
```diff
@@ -543,6 +694,9 @@ grub_linux_boot (void)
 		(unsigned) ctx.real_size,
 		(unsigned) efi_mmap_size);
 
+  if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+    ap_wake_block_size = GRUB_MLE_AP_WAKE_BLOCK_SIZE;
```
Comment (miczyg1) [inline grub-core/loader/i386/linux.c:698]: Few lines lower we need to allocate more memory for the AP wake block:

```
    if (grub_add (ctx.real_size, efi_mmap_size, &sz))
      return GRUB_ERR_OUT_OF_RANGE;
```

change to:

```
    if (grub_add (ctx.real_size, efi_mmap_size + ap_wake_block_size, &sz))
      return GRUB_ERR_OUT_OF_RANGE;
```

File: grub-core/loader/multiboot_elfxx.c [hunk 2]
```diff
@@ -101,30 +106,93 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
...
+      else
+	{
+	  total_size = load_size;
+	  slparams->mle_ptab_size = 0;
+	}
+
+      if (total_size > mld->max_addr || mld->min_addr > mld->max_addr - total_size)
 	return grub_error (GRUB_ERR_BAD_OS, "invalid min/max address and/or load size");
 
-      err = grub_relocator_alloc_chunk_align_safe (GRUB_MULTIBOOT (relocator), &ch,
```
Comment (miczyg1) [inline grub-core/loader/multiboot_elfxx.c:112]: Lets' keep the safe function here, the rest fo changes are ok (i.e. we keep the total_size parameter here)

File: grub-core/loader/multiboot_mbi2.c [hunk 3]
```diff
@@ -301,10 +307,10 @@ grub_multiboot2_load (grub_file_t file, const char *filename)
 	      return grub_error (GRUB_ERR_BAD_OS, "invalid min/max address and/or load size");
 	    }
 
-	  err = grub_relocator_alloc_chunk_align_safe (grub_multiboot2_relocator, &ch,
```
Comment (miczyg1) [inline grub-core/loader/multiboot_mbi2.c:304]: Let's keep the safe function (the `- code_size` is not probably needed). Basically let;s not touch this function call at all.

File: grub-core/loader/multiboot_mbi2.c [hunk 4]
```diff
@@ -714,15 +764,20 @@ grub_multiboot2_make_mbi (grub_uint32_t *target)
   COMPILE_TIME_ASSERT (MULTIBOOT_TAG_ALIGN % sizeof (grub_properly_aligned_t) == 0);
 
   err = grub_relocator_alloc_chunk_align (grub_multiboot2_relocator, &ch,
-					  MBI_MIN_ADDR, UP_TO_TOP32 (bufsize),
+					  MBI_MIN_ADDR, 0xffffffff - bufsize,
```
Comment (miczyg1) [inline grub-core/loader/multiboot_mbi2.c]: Let's not change it. UP_TO_TOP32(value) is probably doing the same as `0xffffffff - value`

**Review by miczyg1 — CHANGES_REQUESTED** [review]
@TomaszAIR Only a few small mistakes to fix

File: grub-core/loader/i386/linux.c (hunk 1 above)
Comment (tomzy-0) [reply grub-core/loader/i386/linux.c:698]: @miczyg1 changed https://github.com/TrenchBoot/grub/pull/8/commits/f228aee64e622d063b08b31941c50b38b97672ee#diff-8c69f5f8f99bf5f4195720ea867e6e8fe273f9261ce151f58222e0d73b34c208R707

File: grub-core/loader/multiboot_elfxx.c (hunk 2 above)
Comment (tomzy-0) [reply grub-core/loader/multiboot_elfxx.c:112]: @miczyg1 changed to safe function https://github.com/TrenchBoot/grub/pull/8/commits/f016297095585e458d2e68fed5e205640775f962#diff-32e991c56790cec87f55280083791c5f9af329a33f9b67cc8998ac8da765c5d4R144

File: grub-core/loader/multiboot_mbi2.c (hunk 3 above)
Comment (tomzy-0) [reply grub-core/loader/multiboot_mbi2.c:304]: @miczyg1 changed to safe, and code_size substraction removed https://github.com/TrenchBoot/grub/pull/8/commits/b3da590ce6b2cece223869954fe3f250ed4628bb#diff-285279252527f5bc911013189ac09a0880271f60dc8f28f28fe68c2f5d517c74R310

File: grub-core/loader/multiboot_mbi2.c (hunk 4 above)
Comment (tomzy-0) [reply grub-core/loader/multiboot_mbi2.c]: @miczyg1 macro used https://github.com/TrenchBoot/grub/pull/8/commits/b3da590ce6b2cece223869954fe3f250ed4628bb#diff-285279252527f5bc911013189ac09a0880271f60dc8f28f28fe68c2f5d517c74R767

**Comment by tomzy-0** [comment]
@miczyg1 I pushed fixes. Now will use those comments to create patches for qubes os

**Review by miczyg1 — APPROVED** [review]

**Pull request merged**